File: Portfolio-app/backend/src/main/java/com/portfolio/pricing/entity/Instrument.java
package com.portfolio.pricing.entity;

import com.portfolio.common.util.AssetClass;
import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDate;

@Entity
@Table(name = "instruments")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Instrument {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private String id;

    @Enumerated(EnumType.STRING)
    @Column(name = "instrument_type", nullable = false)
    private InstrumentType instrumentType;

    @Column(nullable = false)
    private String name;

    private String ticker;

    @Column(name = "exchange_id")
    private String exchangeId;

    @Column(nullable = false, length = 3)
    private String currency;

    @Column(length = 2)
    private String country;

    @Enumerated(EnumType.STRING)
    @Column(name = "asset_class", nullable = false)
    private AssetClass assetClass;

    private String sector;

    private String industry;

    private String provider;

    @Column(name = "expense_ratio", precision = 6, scale = 4)
    private BigDecimal expenseRatio;

    @Column(name = "benchmark_index")
    private String benchmarkIndex;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    @Builder.Default
    private InstrumentStatus status = InstrumentStatus.ACTIVE;

    @Column(name = "first_listed_at")
    private LocalDate firstListedAt;

    @Column(name = "delisted_at")
    private LocalDate delistedAt;

    public enum InstrumentType {
        STOCK,
        ETF,
        ETN,
        BOND,
        COMMODITY_INDEX,
        CASH_PROXY
    }

    public enum InstrumentStatus {
        ACTIVE,
        DELISTED
    }
}

File: frontend/src/api/instrument.ts
import apiClient from './client';

export interface Instrument {
  id: string;
  instrumentType: string;
  name: string;
  ticker: string;
  currency: string;
  country: string;
  assetClass: string;
  sector?: string;
  industry?: string;
  provider?: string;
  status: string;
}

export interface InstrumentSearchParams {
  q?: string;
  assetClass?: string;
  page?: number;
  size?: number;
}

export const instrumentApi = {
  async search(params: InstrumentSearchParams) {
    const response = await apiClient.get('/v1/instruments/search', { params });
    return response.data;
  },

  async getById(id: string) {
    const response = await apiClient.get(`/v1/instruments/${id}`);
    return response.data;
  },

  async getAll(assetClass?: string) {
    const response = await apiClient.get('/v1/instruments', {
      params: { assetClass },
    });
    return response.data;
  },
};
